crates/functional/src/lib.rs

```diff
@@ -0,0 +1,456 @@
+use std::borrow::Borrow;
+use std::cell::RefCell;
+use std::ops::DerefMut;
+use std::rc::Rc;
+use yew::{Component, ComponentLink, Html, Properties};
+
+thread_local! {
+    static CURRENT_HOOK: Rc<RefCell<Option<HookState>>> = Rc::new(RefCell::new(None));
+}
+
+struct HookState {
+    counter: usize,
+    render_call: Rc<dyn Fn()>,
+    hooks: Vec<Rc<RefCell<dyn std::any::Any>>>,
+    next_tick: Vec<Box<dyn FnOnce() -> bool>>,
+}
+
+pub trait FunctionProvider {
+    type TProps: Properties + PartialEq;
+    fn run(props: &Self::TProps) -> Html;
+}
+
+pub struct FunctionComponent<T: FunctionProvider> {
+    _never: std::marker::PhantomData<T>,
+    props: T::TProps,
+    hook_state: RefCell<Option<HookState>>,
+}
+
+impl<T: 'static> Component for FunctionComponent<T>
+where
+    T: FunctionProvider,
+{
+    type Message = ();
+    type Properties = T::TProps;
+
+    fn create(props: Self::Properties, link: ComponentLink<Self>) -> Self {
+        let render_call = Rc::new(move || link.send_message(()));
+        FunctionComponent {
+            _never: std::marker::PhantomData::default(),
+            props,
+            hook_state: RefCell::new(Some(HookState {
+                counter: 0,
+                render_call,
+                hooks: vec![],
+                next_tick: vec![],
+            })),
+        }
+    }
+
+    fn update(&mut self, _: Self::Message) -> bool {
+        let mut should_rerender = false;
+        for scheduled_fn in self
+            .hook_state
+            .borrow_mut()
+            .deref_mut()
+            .as_mut()
+            .expect("Hook state should be mutable here")
+            .next_tick
+            .drain(..)
+        {
+            should_rerender = scheduled_fn() || should_rerender;
+        }
+        should_rerender
```

Comment:
This feels like a step in the wrong direction, yew should be responsible for this optimization

Reply:
You mean if multiple messages are sent there should be only one rerender from yews side?

Reply:
Yeah, exactly. That isn't implemented yet though :)

Reply:
I don't think there is an issue for that yet, is there? :)

Reply:
Nope, would you mind creating one :)

Reply:
https://github.com/yewstack/yew/issues/1034